Feature References audit: configuration (#976)
meilisearch/documentation: reference/features/configuration.md

## Usage

Command-line flags take precedence over environment variables. If an option is specified both as a flag and as an environment variable, MeiliSearch will use the command-line flag and its respective value.

> All options must specify a value. Using a command-line flag or environment variable without specifying a value will throw an error and interrupt the launch process.

Author: ```suggestion
**All configuration options must specify a value.** Using a command-line option or environment variable without specifying a value will throw an error and interrupt the launch process.
```

Command-line flags take precedence over environment variables. If an option is specified both as a flag and an environment variable, the command-line flag and its respective value will be used.
```bash
./meilisearch --schedule-snapshot
```

**Example**:
`meilisearch --schedule-snapshot` throws an error.
`meilisearch --schedule-snapshot=true` activates snapshots scheduling properly.
```bash
error: The argument '--schedule-snapshot <schedule-snapshot>' requires a value but none was supplied
```

## Options

#### General
### General

- [Database path](/reference/features/configuration.md#database-path)
- [Environment](/reference/features/configuration.md#environment)
- [HTTP address & port binding](/reference/features/configuration.md#http-address-port-binding)
- [Master key](/reference/features/configuration.md#master-key)
- [Environment](/reference/features/configuration.md#environment)

#### Advanced
### Advanced

- [Analytics](/reference/features/configuration.md#analytics)
- [Payload Limit Size](/reference/features/configuration.md#payload-limit-size)
- [Snapshots](/reference/features/configuration.md#schedule-snapshot-creation):

>   - [Schedule Snapchot Creation](/reference/features/configuration.md#schedule-snapshot-creation)

Reviewer: RIP little `snapchot`, may you rest well

Author: `snapchot` will always have its spot in my ❤️'s documentation

  - [Snapshot interval](/reference/features/configuration.md#snapshot-interval)
  - [Import snapshot](/reference/features/configuration.md#import-snapshot)
  - [Ignore missing snapshot](/reference/features/configuration.md#ignore-missing-snapshot)
  - [Ignore snapshot if DB exists](/reference/features/configuration.md#ignore-snapshot-if-db-exists)
- [SSL configuration](/reference/features/configuration.md#ssl-authentication-path):
  - [SSL authentication path](/reference/features/configuration.md#ssl-authentication-path)
  - [SSL certificates path](/reference/features/configuration.md#ssl-certificates-path)
  - [SSL key path](/reference/features/configuration.md#ssl-key-path)
  - [SSL OCSP path](/reference/features/configuration.md#ssl-ocsp-path)
  - [SSL require auth](/reference/features/configuration.md#ssl-require-auth)
  - [SSL resumption](/reference/features/configuration.md#ssl-resumption)

>   - [SSL tickets](/reference/features/configuration.md#ssl-tickets)

Author: It was eating me up inside, for real

### Database path

**Environment variable**: `MEILI_DB_PATH`
**CLI option**: `--db-path`

> **Default value**: `"./data.ms"`

Reviewer: I think there is a minor inconsistency in our documentation style here. The default value here is `"./data.ms"`, whereas similar options like snapshot and dump directories have a default value without the `.`, e.g. `snapshots/` and `dumps/`.

Supposing that this is a problem w/ our documentation and not related to how MeiliSearch actually interprets the input, I think we should change this one to match the others:

```suggestion
**Default value**: `data.ms/`
```

Can someone from the core team point to the code where this is implemented? @MarinPostma @ManyTheFish

Sets the location of the database file.

### Environment

**Environment variable**: `MEILI_ENV`
**CLI option**: `--env`
**Default value**: `development`

Configures the instance's environment. Value must be either `production` or `development`.

The main differences between the two environments are:

- `production`: setting a [master key](/reference/features/authentication.md) is **mandatory**. Logging is disabled
- `development`: setting a [master key](/reference/features/authentication.md) is **optional**. Log level is automatically set to `info`

When the server environment is set to `development`, providing a master key is not mandatory. This is useful when debugging and prototyping, but dangerous otherwise since API update routes are unprotected.

Logging and the [web interface](/reference/features/web_interface.md#web-interface) are disabled in `production`.

### HTTP address & port binding

**Environment variable**: `MEILI_HTTP_ADDR`
**CLI option**: `--http-addr`
**Default value**: `"127.0.0.1:7700"`

Sets the HTTP address and port MeiliSearch will use.

### Master key

**Environment variable**: `MEILI_MASTER_KEY`
**CLI option**: `--master-key`

> **Default value**: `None`

Author: Should we add it to every option, then?

> **Default value**: `None`

Author: I think the value of any specific formatting is tied to its consistency. I don't want to add an exception here and I don't think I have the time to go back and add the `expected value` field to every configuration option, so I will solve this in two steps:

1. right now, I will create a warning box explicitly stating this option requires the user to add an alphanumeric string;
2. in the (near?) future, I will add a PR explicitly stating every config option's expected value.

Sets the instance's master key, automatically protecting all routes except `GET /health`. Providing a key is mandatory when `--env` is set to `production`.

If no master key is provided in a `development` environment, all routes will be unprotected and publicly accessible.

If no master key is provided in a `production` environment, MeiliSearch will throw an error and refuse to launch.

[Learn more about MeiliSearch's use of security keys.](/reference/features/authentication.md)

> ### Analytics

Author: Done.

**Environment variable**: `MEILI_NO_ANALYTICS`
**CLI option**: `--no-analytics`
**Default value**: `false`

Deactivates MeiliSearch's built-in analytics and telemetry when set to `true`.

By default, MeiliSearch collects the following data from all instances that do not explicitly opt-out:

- Application version
- Environment (development or production)

> - Number of days since the first start (segment development/production

Author: Oops, sorry about that

- Database size
- Last update time
- Number of updates

> - Number of documents per index

Author: To the best of my knowledge (and according to the current live version of the docs, yes, this is accurate with v0.20.

> If a user does not disable analytics, they may also provide an email address and their server provider by setting the environment variables `MEILI_USER_EMAIL` and `MEILI_SERVER_PROVIDER`.

Author: Yeah, this is a really gnarly bit in pretty much every sense ("not disabled", only exists as an env var, is of questionable value). People can give us their e-mails out of the goodness of their hearts, but I'm not sure we actually do anything with them.

I included these options for the sake of completeness, but looking at it now I wonder if it's not a net win for everyone if we just remove this paragraph and leave the "functionality" undocumented. 

@gmourier, what do you think?

> If a user does not disable analytics, they may also provide an email address and their server provider by setting the environment variables `MEILI_USER_EMAIL` and `MEILI_SERVER_PROVIDER`.

Author: I have removed this paragraph. `MEILI_USER_EMAIL` and `MEILI_SERVER_PROVIDER` add noise for little benefit since none of our users seem to use them (understandably).

All collected data is used solely for the purpose of improving MeiliSearch. A full document describing why we collect this data and how we use it is forthcoming.

### Disable sentry

**Environment variable**: `MEILI_NO_SENTRY`
**CLI option**: `--no-sentry`
**Default value**: `false`

Deactivates Sentry when set to `true`.

We use [Sentry](https://sentry.io) to receive bug reports and diagnostics that help us improve MeiliSearch.

### Dumps destination

**Environment variable**: `MEILI_DUMPS_DIR`
**CLI option**: `--dumps-dir`

> **Default value**: `dumps/`

Author: Will keep this as it is so the option is consistent throughout the page.

**CLI option**: `--max-udb-size`
**Default value**: `107374182400` (100 GiB)

Sets the maximum size of the `update` database. Value must be given in bytes.

The `update` database handles [pending updates](/learn/advanced/asynchronous_updates.md). This is different from the `main` database, which only stores processed data.

The maximum UDB size must be a modulo value of the OS's `PAGE_SIZE`. To find the OS's `PAGE_SIZE`, use the following command:

```bash
getconf PAGE_SIZE
```

On **UNIX** systems (e.g. Linux, MacOS) `getconf` returns the maximum page size.

On **Windows**, `getconf` returns a fixed size that will be allocated when launching the instance.

[Learn more about MeiliSearch's database and storage engine.](/reference/under_the_hood/storage.md)

### Payload limit size

**Environment variable**: `MEILI_HTTP_PAYLOAD_SIZE_LIMIT`
**CLI option**: `--http-payload-size-limit`
**Default value**: `104857600` (~100MB)

Sets the maximum size of accepted JSON payloads. Value must be given in bytes.

### Schedule snapshot creation

**Environment variable**: `MEILI_SCHEDULE_SNAPSHOT`
**CLI option**: `--schedule-snapshot`
**Default value**: `false`

> Activates scheduled snapshots when set to `true`. Disabled by default.

Reviewer: ```suggestion
Activates scheduled snapshots when set to `true`. Snapshots are disabled by default.
```

[Learn more about snapshots](/reference/features/snapshots.md).

### Snapshot destination

**Environment variable**: `MEILI_SNAPSHOT_DIR`
**CLI option**: `--snapshot-dir`
**Default value**: `snapshots/`

Sets the directory where MeiliSearch will store snapshots.

### Snapshot interval

**Environment variable**: `MEILI_SNAPSHOT_INTERVAL_SEC`
**CLI option**: `--snapshot-interval-sec`
**Default value**: `86400` (1 day)

Defines the interval between each snapshot. Value must be given in seconds.

### Import snapshot

**Environment variable**: N/A
**CLI option**: `--import-snapshot`
**Default value**: `None`

> Launches an instance by importing a previously-generated snapshot.

Reviewer: ```suggestion
Launches MeiliSearch after importing a previously-generated snapshot at the given filepath.
```

This command will throw an error if:

- A database already exists

> - No valid snapshot can be found in the specified path

Reviewer: ```suggestion
- No valid snapshot can be found in the specified path

This behavior can be modified with the [`--ignore-snapshot-if-db-exists`](#ignore-snapshot-if-db-exists) and [`--ignore-missing-snapshot`](#ignore-missing-snapshot) options, respectively. 
```

*This option is not available as an environment variable.*

### Ignore missing snapshot

**Environment variable**: N/A
**CLI option**: `--ignore-missing-snapshot`
**Default value**: `false`

> Prevents a MeiliSearch instance from throwing an error when `--import-snapshot` does not point to a valid snapshot file.

Reviewer: ```suggestion
Prevents a MeiliSearch instance from throwing an error when [`--import-snapshot`](#import-snapshot) does not point to a valid snapshot file.
```

This command will throw an error if `--import-snapshot` is not defined.

*This option is not available as an environment variable.*

### Ignore snapshot if DB exists

**Environment variable**: N/A
**CLI option**: `--ignore-snapshot-if-db-exists`
**Default value**: `false`

> Prevents a MeiliSearch instance with an existing database from throwing an error when using `--import-snapshot`.

Reviewer: ```suggestion
Prevents a MeiliSearch instance with an existing database from throwing an error when using `--import-snapshot`. Instead, the snapshot will be ignored and MeiliSearch will launch using the existing database.
```
